File: ui-testing/pages/login_page.py
from pages.base_page import BasePage
from selenium.webdriver.common.by import By


class LoginPage(BasePage):

    URL = 'https://www.saucedemo.com/'
    # ============================================================
    # (Actively Used Locators)
    # ============================================================
    USER_ACCOUNT = (By.ID, 'user-name')
    PASSWORD_INPUT = (By.ID, 'password')
    LOGIN_BUTTON = (By.ID, 'login-button')
    ERROR_CONTAINER = (By.CSS_SELECTOR, "h3[data-test='error']")
    CLOSE_REMIND = (By.CLASS_NAME, 'error-button')

    # ============================================================
    # (Reserved / Pending Cleanup)
    # ============================================================
    LOGO = (By.CLASS_NAME, 'login_logo')
    All_USERNAMES = (By.XPATH, '//*[@id="login_credentials"]')
    All_PASSWORD = (By.XPATH, '//*[@id="root"]/div/div[2]/div[2]/div/div[2]')
    ACCEPTED_USERNAMES = 'standard_user'
    ACCEPTED_PASSWORD = 'secret_sauce'
    ERROR_REMIND = (By.XPATH, '//*[@id="login_button_container"]/div/form/div[3]/h3')

    def login(self, username, password):
        self.driver.get(self.URL)
        self.fill(self.USER_ACCOUNT, username)
        self.fill(self.PASSWORD_INPUT, password)
        self.click(self.LOGIN_BUTTON)

    def get_error_message(self):
        return self.get_text(self.ERROR_CONTAINER)

    def click_error_button(self):
        self.click(self.CLOSE_REMIND)

    def is_error_message_hidden(self) -> bool:
        return not self.is_visible(self.ERROR_CONTAINER)

File: playwright-testing/pages/login_page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base.page';

export class LoginPage extends BasePage {
    readonly url = 'https://www.saucedemo.com/';

    readonly usernameInput: Locator;
    readonly passwordInput: Locator;
    readonly loginButton: Locator;
    readonly errorContainer: Locator;
    readonly closeErrorButton: Locator;

    readonly logo: Locator;
    readonly allUsernames: Locator;
    readonly allPassword: Locator;

    readonly acceptedUsername = 'standard_user';
    readonly acceptedPassword = 'secret_sauce';

    constructor(page: Page) {
        super(page);
        this.usernameInput = page.locator('#user-name');
        this.passwordInput = page.locator('#password');
        this.loginButton = page.locator('#login-button');
        this.errorContainer = page.locator('[data-test="error"]');
        this.closeErrorButton = page.locator('.error-button');

        this.logo = page.locator('.login_logo');
        this.allUsernames = page.locator('#login_credentials');
        this.allPassword = page.locator('.login_password');
    }
    async goto(): Promise<void> {
        await this.navigate(this.url);
    }

    async login(username: string, password: string): Promise<void> {
        await this.usernameInput.fill(username);
        await this.passwordInput.fill(password);
        await this.loginButton.click();
    }

    async getErrorMessage(): Promise<string> {
        return await this.errorContainer.innerText();
    }


    async clickErrorButton(): Promise<void> {
        await this.closeErrorButton.click();
    }
}